### PR TITLE
opt(assignment): speedup EndBlock padding

### PR DESCRIPTION
**Pull request by lispc**
in evm circuit, if padding is enabled, then a lot of EndBlock will be assigned with the expensive `CachedRegion` way.

In this PR, these same non-last EndBlock steps are only handled by CachedRegion once, and then are repeated until the last EndBlock. 

For a 1<<18 circuit, assignment time reduced from 28s to 4s in my laptop

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util.rs [hunk 1]
```diff
@@ -84,20 +86,52 @@ impl<'r, 'b, F: FieldExt> CachedRegion<'r, 'b, F> {
...
+            width_start: advice_columns[0].index(),
             height_start,
+            advice_columns,
         }
     }
 
+    /// Repeatedly assign the locally cached witnesses.
+    /// This method can be used as a "quick" path for assignment for repeated
+    /// padding rows
+    pub fn replicate_assignment_for_range<A, AR>(
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/util.rs:106]: I would suggest the following small change to this function comment in order to make it more clear what it does:

Add to the comment indicating that this function replicates the assignment of 1 row at `height_start` (which must be already assigned via the `CachedRegion`) into a range of rows indicated by `offset_begin, offset_end`.

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for working on this improvement!
I've added two comments with suggestions for improvements/clarity.

File: zkevm-circuits/src/evm_circuit/execution.rs [hunk 2]
```diff
@@ -774,46 +775,56 @@ impl<F: Field> ExecutionConfig<F> {
...
-                                evm_rows
-                            );
-                            return Err(Error::Synthesis);
-                        }
-                        if exact || evm_rows - (offset + cur_state_height) == 1 {
-                            block_step = end_block_last;
-                            no_next_step = true;
+                let no_padding = evm_rows == 0;
+
+                let assign_q_step =
```
[the remark this answers is not in the document]
Comment (lispc) [reply zkevm-circuits/src/evm_circuit/execution.rs]: done

File: zkevm-circuits/src/evm_circuit/util.rs (hunk 1 above)
Comment (lispc) [reply zkevm-circuits/src/evm_circuit/util.rs:107]: done

**Review by han0110 — APPROVED** [review]
LGTM! Nice work!

**Pull request merged**